dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java

```diff
@@ -47,120 +55,189 @@ public UnitOfWorkAspect(Map<String, SessionFactory> sessionFactories) {
+                List<String> unregistered = Arrays.asList(unitOfWork.value()).stream()
+                    .filter(name -> !sessionFactories.containsKey(name))
+                    .collect(Collectors.toList());
+
+                throw new IllegalArgumentException("Unregistered Hibernate bundle/s: '" + String.join("','", unregistered) + "'");
             }
         }
-        session = sessionFactory.openSession();
+
+        sessions = selectedSessionFactories.entrySet().stream()
+            .map(entry -> Pair.of(entry.getKey(), entry.getValue().openSession()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
         try {
-            configureSession();
-            ManagedSessionContext.bind(session);
-            beginTransaction(unitOfWork, session);
+            configureSessions();
+            sessions.values().forEach(ManagedSessionContext::bind);
+            beginTransactions(unitOfWork, sessions);
         } catch (Throwable th) {
-            session.close();
-            session = null;
-            ManagedSessionContext.unbind(sessionFactory);
+            onFinish();
             throw th;
         }
     }
 
     public void afterEnd() {
-        if (unitOfWork == null || session == null) {
+        if (unitOfWork == null || sessions.isEmpty()) {
             return;
         }
 
         try {
-            commitTransaction(unitOfWork, session);
+            commitTransactions(unitOfWork, sessions);
         } catch (Exception e) {
-            rollbackTransaction(unitOfWork, session);
+            rollbackTransactions(unitOfWork, sessions);
             throw e;
         }
-        // We should not close the session to let the lazy loading work during serializing a response to the client.
-        // If the response successfully serialized, then the session will be closed by the `onFinish` method
+        // We should not close the sessions to let the lazy loading work during serializing a response to the client.
+        // If the response successfully serialized, then the sessions will be closed by the `onFinish` method
     }
 
     public void onError() {
-        if (unitOfWork == null || session == null) {
+        if (unitOfWork == null || sessions.isEmpty()) {
             return;
         }
 
         try {
-            rollbackTransaction(unitOfWork, session);
+            rollbackTransactions(unitOfWork, sessions);
         } finally {
             onFinish();
         }
     }
 
     public void onFinish() {
         try {
-            if (session != null) {
-                session.close();
+            //We need to ensure that all the sessions have been closed
+            //and also we need to inform about the exceptions
+            List<Optional<RuntimeException>> errors = sessions.values().stream()
+                .map(UnitOfWorkAspect::closeSession)
+                .filter(Optional::isPresent)
```

Comment:
Bit of a nitpick, but it might be a little cleaner if you call `.map(Optional::get)`, since you’re already filtering by `isPresent` here.